src/main/java/hudson/plugins/ec2/EC2HostAddressProvider.java

```diff
@@ -0,0 +1,97 @@
+package hudson.plugins.ec2;
+
+import static hudson.plugins.ec2.ConnectionStrategy.PRIVATE_DNS;
+import static hudson.plugins.ec2.ConnectionStrategy.PRIVATE_IP;
+import static hudson.plugins.ec2.ConnectionStrategy.PUBLIC_DNS;
+import static hudson.plugins.ec2.ConnectionStrategy.PUBLIC_IP;
+
+import com.amazonaws.services.ec2.model.Instance;
+import com.google.common.base.Objects;
+import org.apache.commons.lang.StringUtils;
+
+import java.util.Optional;
+
+public class EC2HostAddressProvider {
+    public static String unix(Instance instance, ConnectionStrategy strategy) {
+        switch (strategy) {
+            case PUBLIC_DNS:
+                return filterNonEmpty(getPublicDnsName(instance)).orElse(getPublicIpAddress(instance));
+            case PUBLIC_IP:
+                return getPublicIpAddress(instance);
+            case PRIVATE_DNS:
+                return filterNonEmpty(getPrivateDnsName(instance)).orElse(getPrivateIpAddress(instance));
+            case PRIVATE_IP:
+                return getPrivateIpAddress(instance);
+            default:
+                throw new IllegalArgumentException("Could not unix host address for strategy = " + strategy.toString());
+        }
+    }
+
+    public static HostWithIpAddress windows(Instance instance, ConnectionStrategy strategy) {
+        if (strategy.equals(PRIVATE_DNS) || strategy.equals(PRIVATE_IP)) {
+            String dns = filterNonEmpty(getPrivateDnsName(instance)).orElse("");
+            return new HostWithIpAddress(dns, getPrivateIpAddress(instance));
+        } else if (strategy.equals(PUBLIC_DNS) || strategy.equals(PUBLIC_IP)) {
+            return filterNonEmpty(getPublicDnsName(instance))
+                    .map(dns -> new HostWithIpAddress(dns, getPublicIpAddress(instance)))
+                    .orElseGet(() -> {
+                        String dns = filterNonEmpty(getPrivateDnsName(instance)).orElse("");
+                        return new HostWithIpAddress(dns, getPrivateIpAddress(instance));
+                    });
+        } else {
+            throw new IllegalArgumentException("Could not unix host address for strategy = " + strategy.toString());
+        }
+    }
+
+    private static String getPublicDnsName(Instance instance) {
+        return instance.getPublicDnsName();
+    }
+
+    private static String getPublicIpAddress(Instance instance) {
+        return instance.getPublicIpAddress();
+    }
+
+    private static String getPrivateDnsName(Instance instance) {
+        return instance.getPrivateDnsName();
+    }
+
+    private static String getPrivateIpAddress(Instance instance) {
+        return instance.getPrivateIpAddress();
+    }
+
+    private static Optional<String> filterNonEmpty(String value) {
+        return Optional.ofNullable(value).filter(StringUtils::isNotEmpty);
+    }
+
+    public static class HostWithIpAddress {
```

Comment:
You could potentially use `com.google.common.net.HostAndPort` here, or even `java.net.InetSocketAddress` if equality isn't too important (InetSocketAddress does some actual DNS lookups and such to compare different address/port combos).